libfoundation/src/foundation-objc.mm

```diff
@@ -383,6 +383,809 @@ MC_DLLEXPORT_DEF uintptr_t MCObjcObjectGetActionProxySelector(void)
+{
+    MCAutoStringRefAsCString t_protocol_cstring;
+    if (!t_protocol_cstring.Lock(p_protocol_name))
+        return false;
+    
+    // Get protocol object
+    Protocol *t_protocol = objc_getProtocol(*t_protocol_cstring);
+    if (t_protocol == nullptr)
+    {
+        return MCErrorCreateAndThrowWithMessage(kMCObjcDelegateCallbackSignatureErrorTypeInfo,
+                                                MCSTR("Protocol %{protocol} not found"),
+                                                "protocol", p_protocol_name,
+                                                nullptr);
+    }
+    
+    MCNameRef t_key;
+    MCValueRef t_value;
+    uintptr_t t_iterator = 0;
+    while (MCArrayIterate(p_handler_mapping, t_iterator, t_key, t_value))
+    {
+        if (!__HandlerConformsToProtocolMethod(static_cast<MCHandlerRef>(t_value),
+                                               MCNameGetString(t_key),
+                                               p_context,
+                                               t_protocol))
+            return false;
+    }
+    
+    com_livecode_MCObjcFormalDelegate *t_proxy =
+    [[com_livecode_MCObjcFormalDelegate alloc] initForProtocol: t_protocol withHandlerMapping:p_handler_mapping withContext:p_context];
+    
+    MCObjcObjectRef t_proxy_object = nullptr;
+    if (!MCObjcObjectCreateWithRetainedId(t_proxy, t_proxy_object))
+    {
+        [t_proxy release];
+        return false;
+    }
+    
+    r_object = t_proxy_object;
+    return true;
+}
+
+extern "C" MC_DLLEXPORT_DEF
+bool MCObjcCreateDelegate(MCStringRef p_protocol_name, MCArrayRef p_handler_mapping, MCObjcObjectRef& r_object)
+{
+    return MCObjcCreateDelegateWithContext(p_protocol_name,
+                                           p_handler_mapping,
+                                           nullptr,
+                                           r_object);
+}
+
+extern "C" MC_DLLEXPORT_DEF
+bool MCObjcCreateInformalDelegateWithContext(MCProperListRef p_foreign_handlers, MCArrayRef p_handler_mapping, MCValueRef p_context, MCObjcObjectRef& r_object)
+{
+    MCAutoArrayRef t_selector_mapping;
+    if (!MCArrayCreateMutable(&t_selector_mapping))
+        return false;
+    
+    for (uindex_t i = 0; i < MCProperListGetLength(p_foreign_handlers); ++i)
+    {
+        MCValueRef t_element =
+            MCProperListFetchElementAtIndex(p_foreign_handlers, i);
+        
+        if (MCValueGetTypeCode(t_element) != kMCValueTypeCodeHandler)
+            return false;
+        
+        MCHandlerRef t_handler = static_cast<MCHandlerRef>(t_element);
+        
+        const MCHandlerCallbacks *t_callbacks = MCHandlerGetCallbacks(t_handler);
+        if (t_callbacks->query == nullptr)
+            return false;
+        
+        SEL t_selector = nullptr;
+        t_callbacks->query(MCHandlerGetContext(t_handler),
```

Comment:
You need to check the return value is not false here - in case someone tries to pass a non-obc-foreign handler!